File: src/main/java/cspom/compiler/patterns/DeReify.java
package cspom.compiler.patterns;

import java.util.Deque;

import cspom.CSPOM;
import cspom.constraint.CSPOMConstraint;
import cspom.constraint.FunctionalConstraint;
import cspom.constraint.GeneralConstraint;
import cspom.variable.TrueDomain$;

public final class DeReify implements ConstraintCompiler {

    private final CSPOM problem;
    private final Deque<CSPOMConstraint> constraints;

    public DeReify(final CSPOM problem, final Deque<CSPOMConstraint> constraints) {
        this.problem = problem;
        this.constraints = constraints;
    }

    @Override
    public void compile(final CSPOMConstraint c) {
        if (c instanceof FunctionalConstraint) {
            final FunctionalConstraint fc = (FunctionalConstraint) c;
            if (TrueDomain$.MODULE$ == fc.result().domain()) {
                deReify(fc);
            }
        }
    }

    private void deReify(final FunctionalConstraint constraint) {
        problem.removeConstraint(constraint);
        final CSPOMConstraint newConstraint = new GeneralConstraint(
                constraint.description(), constraint.parameters(),
                constraint.arguments());
        problem.addConstraint(newConstraint);
        constraints.add(newConstraint);
    }

}

File: src/main/scala/cspom/compiler/patterns/DeReify.scala
package cspom.compiler.patterns;

import cspom.constraint.{ GeneralConstraint, FunctionalConstraint, CSPOMConstraint }
import cspom.variable.TrueDomain
import cspom.CSPOM
import scala.collection.mutable.Queue
import cspom.constraint.Predicate

final class DeReify(
  private val problem: CSPOM,
  private val constraints: Queue[CSPOMConstraint]) extends ConstraintCompiler {

  override def compileFunctional(fc: FunctionalConstraint) = {
    if (fc.result.domainOption.map(_ == TrueDomain).getOrElse(false)) {
      problem.removeConstraint(fc);
      val newConstraint = new GeneralConstraint(
        Predicate(fc.predicate.function, fc.predicate.parameters), fc.arguments);
      problem.addConstraint(newConstraint);
      // constraints.enqueue(newConstraint);
      true
    } else false

  }

}
